### PR TITLE
[FSDP] Use _init_from_local_tensor to create ShardedTensor to avoid communication overhead

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77628
* __->__ #82911

FSDP originally uses `_init_from_local_shards_and_global_metadata()` to create a ShardedTensor for sharded_state_dict(). We have seen some non-trivial overhead if the number of tensors is large. Using `_init_from_local_shards_and_global_metadata ` can significantly reduce the overhead. For a model with ~250 tensors in the state_dict trained with 16 GPUs, the original `sharded_state_dict` takes ~1.7 seconds and this PR reduces the overhead to ~0.6 seconds.

Differential Revision: [D38452170](https://our.internmc.facebook.com/intern/diff/D38452170/)

### CODE REVIEW

File: torch/distributed/fsdp/shard_utils.py [hunk 1]
```diff
@@ -188,3 +194,56 @@ def _gather_state_dict(
...
+    chunks = tensor.chunk(world_size)
+    if len(chunks) > rank:
+        local_shard = chunks[rank].clone()
+        offsets = [0 for _ in tensor.size()]
+        offsets[0] = math.ceil(tensor.size()[0] / world_size) * rank
+        local_shards = [Shard.from_tensor_and_offsets(local_shard, offsets, rank)]
+    else:
+        local_shards = []
+
+    # Create a ShardedTensor without invoking communnication.
```
Comment (awgu) [inline torch/distributed/fsdp/shard_utils.py:219]: nit:
```suggestion
    # Create a ShardedTensor without invoking communication.
```

File: torch/distributed/fsdp/shard_utils.py [hunk 2]
```diff
@@ -188,3 +194,56 @@ def _gather_state_dict(
...
+        size=tensor.size(),
+        tensor_properties=TensorProperties(
+            dtype=tensor.dtype,
+            layout=tensor.layout,
+            requires_grad=False,
+            memory_format=torch.contiguous_format,
+            pin_memory=tensor.is_pinned(),
+        )
+    )
+    return ShardedTensor._init_from_local_shards_and_global_metadata(
```
Comment (awgu) [inline torch/distributed/fsdp/shard_utils.py:245]: It looks like this does not use `_init_from_local_tensor()` (anymore)? Maybe we should update the PR title/summary.

File: torch/distributed/fsdp/shard_utils.py [hunk 3]
```diff
@@ -188,3 +194,56 @@ def _gather_state_dict(
...
+    rank: int,
+    world_size: int,
+    device_per_node: int,
+    pg: dist.ProcessGroup,
+) -> ShardedTensor:
+    """
+    Shard a tensor to chunks along the first dimension. The local rank will gets its
+    corresponding chunk as the local shard to create a ShardedTensor.
+    """
+    chunks = tensor.chunk(world_size)
```
Comment (awgu) [inline torch/distributed/fsdp/shard_utils.py:210]: nit: Do you think it is worth explicitly passing `dim=0` into `chunk()`? It might make it more clear that `chunk()` is chunking along the 0th dimension.

File: torch/distributed/fsdp/shard_utils.py [hunk 4]
```diff
@@ -188,3 +194,56 @@ def _gather_state_dict(
...
+    if len(chunks) > rank:
+        local_shard = chunks[rank].clone()
+        offsets = [0 for _ in tensor.size()]
+        offsets[0] = math.ceil(tensor.size()[0] / world_size) * rank
+        local_shards = [Shard.from_tensor_and_offsets(local_shard, offsets, rank)]
```
Comment (awgu) [inline torch/distributed/fsdp/shard_utils.py:215]: Conceptual check: `local_shards` here has length 1? (So overall, it is either length 1 if this rank has a chunk, or length 0 if this rank does not have a chunk?)

File: test/distributed/fsdp/test_shard_utils.py [hunk 5]
```diff
@@ -185,3 +186,20 @@ def get_offsets(tensor, shard):
         uneven_sharded_tensor.gather(0, output)
         if self.rank == 0:
             self.assertEqual(tensor, output)
+
+    @skip_if_lt_x_gpu(2)
+    def test_create_chunk_sharded_tensor(self):
+        for size in ((1,), (1, 6), (12,), (12, 6), (25,), (25, 6)):
```
Comment (awgu) [inline test/distributed/fsdp/test_shard_utils.py:192]: Do you think testing at most 2 dimensions suffices? I guess it should be fine since you are only sharding along dim 0 so as long as there is a second dim, it should be okay.

**Review by awgu — APPROVED** [review]
LGTM. I left some nits, but I am stamping to unblock.

File: torch/distributed/fsdp/shard_utils.py (hunk 4 above)
Comment (fegin) [reply torch/distributed/fsdp/shard_utils.py:215]: That's correct!

File: test/distributed/fsdp/test_shard_utils.py (hunk 5 above)
Comment (fegin) [reply test/distributed/fsdp/test_shard_utils.py:192]: Unless we change the implementation to chunk in a different dimension, 2 dimensions test should be enough.
